program/tests/transfer_hook.rs

```diff
@@ -0,0 +1,1349 @@
+//! Most of these test cases are checked by Token-2022, but it doesn't hurt to
+//! check them by directly invoking the program's `ExecuteInstruction`.
+
+#![cfg(feature = "test-sbf")]
+
+mod setup;
+
+use {
+    paladin_rewards_program::{
+        error::PaladinRewardsError,
+        state::{get_holder_rewards_address, get_holder_rewards_pool_address, HolderRewards},
+    },
+    setup::{
+        setup, setup_extra_metas_account, setup_holder_rewards_account,
+        setup_holder_rewards_pool_account, setup_mint, setup_token_account,
+        setup_token_account_transferring,
+    },
+    solana_program_test::*,
+    solana_sdk::{
+        account::AccountSharedData,
+        instruction::{AccountMeta, Instruction, InstructionError},
+        pubkey::Pubkey,
+        signature::Keypair,
+        signer::Signer,
+        transaction::{Transaction, TransactionError},
+    },
+    spl_associated_token_account::get_associated_token_address,
+    spl_transfer_hook_interface::{
+        error::TransferHookError, get_extra_account_metas_address,
+        offchain::add_extra_account_metas_for_execute,
+    },
+    test_case::test_case,
+};
+
+#[allow(clippy::too_many_arguments)]
+fn execute_with_extra_metas_instruction(
+    source: &Pubkey,
+    mint: &Pubkey,
+    destination: &Pubkey,
+    owner: &Pubkey,
+    holder_rewards_pool: &Pubkey,
+    source_holder_rewards: &Pubkey,
+    destination_holder_rewards: &Pubkey,
+    amount: u64,
+) -> Instruction {
+    spl_transfer_hook_interface::instruction::execute_with_extra_account_metas(
+        &paladin_rewards_program::id(),
+        source,
+        mint,
+        destination,
+        owner,
+        &Pubkey::new_unique(), // (Extra metas) Doesn't matter if we're invoking directly.
+        &[
+            AccountMeta::new_readonly(*holder_rewards_pool, false),
+            AccountMeta::new(*source_holder_rewards, false),
+            AccountMeta::new(*destination_holder_rewards, false),
+        ],
+        amount,
+    )
+}
+
+async fn transfer_with_extra_metas_instruction(
+    context: &mut ProgramTestContext,
+    source: &Pubkey,
+    mint: &Pubkey,
+    destination: &Pubkey,
+    owner: &Pubkey,
+    amount: u64,
+    decimals: u8,
+) -> Instruction {
+    let mut instruction = spl_token_2022::instruction::transfer_checked(
+        &spl_token_2022::id(),
+        source,
+        mint,
+        destination,
+        owner,
+        &[],
+        amount,
+        decimals,
+    )
+    .unwrap();
+
+    // The closure required by `add_extra_account_metas_for_execute` is a pain,
+    // so just grab the extra metas account ahead of time, since we know our
+    // extra metas don't require account data, therefore don't require loading
+    // any other accounts.
```

Comment:
Totally unrelated: we could consider adding some other helper in token-2022 to create the closure given a client and maybe make this easier

Reply:
Yeah, we probably should. It's kind of rough to use on its own, mostly because of lifetimes and async stuff.

Reply:
I think that's why I kept it as limited as it is :sweat_smile: